File: src/FirstBadVersion.java
public class FirstBadVersion {
    public int firstBadVersion(int n) {
        int left = 1, right = n, mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (isBadVersion(mid) && !isBadVersion(mid - 1)) return mid;
            else if (!isBadVersion(mid)) right = mid - 1;
            else left = mid + 1;
        }
        return left;
    }

    // implemented by leetcode (just reference here)
    private static boolean isBadVersion(int version) {
        return false;
    }
}

File: python/first_bad_version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
def isBadVersion(version):
    return None


class Solution:
    def firstBadVersion(self, n: int) -> int:
        """
        :type n: int
        :rtype: int
        """
        left, right = 0, n
        while left <= right:
            middle = left + (right - left) // 2
            if isBadVersion(middle):
                right = middle - 1
            else:
                left = middle + 1
        return left
